rfd/0234-in-band-mfa-ssh-sessions.md

```diff
@@ -0,0 +1,767 @@
+   with the per-session MFA certificate. This complexity increases the risk of implementation errors and inconsistencies
+   across different clients (e.g., `tsh`, web terminal, Teleport Connect, etc.).
+1. A per-session MFA certificate is a single credential, representing multiple factors of authentication for a user. If
+   an attacker were to gain possession, it could be used to bypass all forms of authentication checks, including MFA.
+
+By moving MFA enforcement to the SSH service during session establishment, this new design directly addresses the above
+issues by:
+
+1. MFA enforcement is tightly integrated with session creation (i.e., in-band), ensuring that authentication factors are
+   directly bound to each session and mitigating the risk of bypasses like those seen in
+   [CVE-2025-49825](https://github.com/gravitational/Teleport/security/advisories/GHSA-8cqv-pj7f-pwpc).
+1. MFA complexity is moved from clients to the SSH service, so clients only need to connect to the SSH service and
+   resolve an MFA challenge only if required, streamlining the session establishment process and reducing the risk of
+   implementation errors.
+1. Per-session MFA certificates can be completely removed, eliminating a single credential representing multiple factors
+   of authentication.
+
+In summary, this RFD proposes a more secure and streamlined approach to MFA enforcement for SSH sessions by integrating
+MFA checks directly into the session establishment process, reducing client complexity, and eliminating the need for
+per-session MFA certificates.
+
+## Non-Goals
+
+1. This RFD does not propose changes to other Teleport access protocols such as Kubernetes, databases, desktops, etc.
+   However, the architecture could be extended to these protocols in the future.
+
+## Details
+
+### UX
+
+No changes are expected since this is an internal change.
+
+### High-Level Flow
+
+The client will first dial its target SSH host using the Proxy and its standard Teleport client certificate. The Proxy
+will then resolve the target cluster and host, and invoke the `EvaluateSSHAccess` RPC of the [Decision
+service](https://github.com/gravitational/Teleport.e/blob/master/rfd/0024e-access-control-decision-api.md). Upon
+receiving a permit from the Decision service, the Proxy will staple the permit to the connection and open a connection
+to the respective cluster's SSH service.
+
+The SSH service will authenticate the client using the provided client certificate and the stapled permit. The SSH
+service will then check if MFA is required for the session by examining the permit.
+
+If MFA is _not required_, the SSH service will then proceed to establish the SSH session.
+
+If MFA _is required_, the SSH service will send a Protobuf [`MFAPrompt`
+message](#ssh-keyboard-interactive-authentication) via the SSH [keyboard-interactive
+channel](https://www.rfc-editor.org/rfc/rfc4256) to inform the client that MFA is needed. The client must then call the
```

Comment:
`keyboard-interactive` is not a channel, it's an authentication method.

Reply:
The reason I objected to the wording is that "channel" is a technical term in SSH connections; a "channel" being used for MFA seems to suggest that it's something we're doing after authentication in a SSH channel.